<script type="text/javascript">

    Ext4.onReady(function () {

        /**
         * @cfg outputFileIds
         */
        Ext4.define('VariantDB.panel.ImputationPanel', {
            extend: 'Ext.form.Panel',

            initComponent: function(){
                Ext4.QuickTips.init();

                Ext4.apply(this, {
                    border: false,
                    defaults: {
                        bodyStyle: 'padding: 5px;',
                        style: 'padding-bottom: 10px;',
                        fieldDefaults: {
                            width: 400,
                            labelWidth: 180
                        }
                    },
                    items: [{
                        html: 'Loading...',
                        border: false
                    }],
                    buttonAlign: 'left',
                    buttons: [{
                        text: 'Submit',
                        itemId: 'submitBtn',
                        disabled: true,
                        scope: this,
                        handler: this.onSubmit
                    }]
                });

                this.callParent(arguments);
                this.loadSampleInfo();
            },

            loadSampleInfo: function(){
                Ext4.Msg.wait('Loading sample information...');

                LABKEY.Ajax.request({
                    url: LABKEY.ActionURL.buildURL('variantdb', 'getSamplesFromVcf', null, {outputFileIds: this.outputFileIds}),
                    failure: LDK.Utils.getErrorCallback(),
                    success: LABKEY.Utils.getCallbackWrapper(this.onLoad, this, false),
                    scope: this
                })
            },

            onLoad: function(results){
                Ext4.Msg.hide();

                this.sampleNames = [];
                this.sampleStore = Ext4.create('Ext.data.Store', {
                    fields: ['outputFileId', 'sampleName', 'fileName', 'sampleId']
                });

                this.imputationStore = Ext4.create('Ext.data.Store', {
                    fields: ['outputFileId', 'sampleName', 'fileName', 'sampleId']
                });

                this.fileStore = Ext4.create('Ext.data.Store', {
                    fields: ['outputFileId', 'fileName', 'dataId']
                });

                var outputFileMap = results.outputFileMap || {};
                for (var rowId in results.samples){
                    Ext4.Array.forEach(results.samples[rowId], function(sampleName){
                        this.sampleStore.add(this.sampleStore.createModel({
                            fileName: outputFileMap[rowId].fileName,
                            sampleName: sampleName,
                            outputFileId: rowId,
                            sampleId: rowId + '||' + sampleName
                        }));

                        this.imputationStore.add(this.imputationStore.createModel({
                            fileName: outputFileMap[rowId].fileName,
                            sampleName: sampleName,
                            outputFileId: rowId,
                            sampleId: rowId + '||' + sampleName
                        }));

                        this.sampleNames.push(sampleName);
                    }, this);
                }
                this.sampleNames = Ext4.unique(this.sampleNames);

                var genomeIds = [];
                for (var outputFileId in outputFileMap){
                    if (outputFileMap[outputFileId].genomeId){
                        genomeIds.push(outputFileMap[outputFileId].genomeId);
                    }

                    this.fileStore.add({outputFileId: outputFileId, dataId: outputFileMap[outputFileId].dataId, fileName: outputFileMap[outputFileId].fileName});
                }

                genomeIds = Ext4.unique(genomeIds);
                if (genomeIds.length != 1){
                    Ext4.Msg.alert('Error', 'All files must use the same base genome.  Please select different files');
                    return;
                }

                this.genomeId = genomeIds[0];

                this.addItems();
                this.down('#submitBtn').setDisabled(false);
            },

            addItems: function(){
                this.removeAll();
                this.add([{
                    xtype: 'panel',
                    defaults: {
                        border: false
                    },
                    items: [{
                        title: 'Instructions',
                        html: 'This allows you to perform genotype imputation on the selected VCF files.  This is not currently designed to be a very end-user friendly UI.  It will capture parameters, most of which are passed directly to the tools that are run.  For each field, hover over the question mark for more information.  You will need to provide 2 VCF files, one with the framework marker genotypes and one with dense markers.  You can then choose any number of combinations of samples to run, choosing which animals to select from each VCF.',
                        style: 'padding-bottom: 10px;'
                    },{
                        xtype: 'textfield',
                        fieldLabel: 'Job Name',
                        itemId: 'jobName',
                        width: 400,
                        value: null
                    }]
                },{
                    title: 'Samples',
                    items: [{
                        xtype: 'dataview',
                        store: this.sampleStore,
                        tpl: [
                            '<table class="fileNames"><tr class="fileNames"><td>File Id</td><td style="min-width: 200px;">Sample Name</td><td>Sample Id</td><td>File Name</td></tr>',
                            '<tpl for=".">',
                            '<tr class="file_list">',
                            '<td>{outputFileId:htmlEncode}</td>',
                            '<td>{sampleName:htmlEncode}</td>',
                            '<td>{sampleId:htmlEncode}</td>',
                            '<td>{fileName:htmlEncode}</td>',
                            '</tr>',
                            '</tpl>',
                            '</table>'
                        ]
                    }]
                },{
                    title: 'Step 1: Markers',
                    minHeight: 200,
                    xtype: 'form',
                    defaults: {
                        border: false
                    },
                    items: [{
                        xtype: 'ldk-expdatafield',
                        fieldLabel: 'Framework Markers File',
                        allowBlank: false,
                        helpPopup: 'This should be a BED file containing the coordinates to use as framework markers',
                        name: 'frameworkFile',
                        value: LABKEY.ActionURL.getParameter('frameworkFile')
                    },{
                        xtype: 'ldk-expdatafield',
                        fieldLabel: 'Dense Markers File',
                        allowBlank: false,
                        helpPopup: 'This should be a BED file containing the coordinates to use as dense markers',
                        name: 'denseFile',
                        value: LABKEY.ActionURL.getParameter('denseFile')
                    },{
                        xtype: 'ldk-expdatafield',
                        fieldLabel: 'Allele Frequency File',
                        allowBlank: false,
                        helpPopup: 'This should be the file used for allele frequencies.  This VCF does not need to contain your sample data.  This will be parsed from the AF tag',
                        name: 'alleleFrequencyFile',
                        value: LABKEY.ActionURL.getParameter('alleleFrequencyFile')
                    },{
                        xtype: 'ldk-expdatafield',
                        fieldLabel: 'Interval Blacklist File',
                        allowBlank: false,
                        helpPopup: 'This should be a BED file containing a list of intervals to blacklist.  For each, the genotypes will be set to no call; however, these markers will be retained in the output.',
                        name: 'blacklistFile',
                        value: LABKEY.ActionURL.getParameter('blacklistFile')
                    }]
                },{
                    title: 'Step 2: Pedigree',
                    defaults: {
                        border: false
                    },
                    items: [{
                        html: 'This step uses pedigree information stored in the Laboratory module\'s subjects table.  Please review the information below.  If it is not correct, imputation will not work properly.',
                        style: 'padding-bottom: 10px;'
                    },{
                        xtype: 'ldk-querypanel',
                        border: true,
                        queryConfig: {
                            containerPath: Laboratory.Utils.getQueryContainerPath(),
                            schemaName: 'laboratory',
                            queryName: 'subjects',
                            filterArray: [LABKEY.Filter.create('subjectname', this.sampleNames.join(';'), LABKEY.Filter.Types.IN)]
                        }
                    }]
                },{
                    title: 'Step 3: Imputation Configuration',
                    minHeight: 500,
                    defaults: {
                        border: false
                    },
                    items: [{
                        html: 'For imputation, some of the samples should have comprehensive genotypes, while some should have sparse genotyping.  Use the grid below to separate which samples are complete.  Note: this step is largely added for the purposes of testing.  Once complete, each target will be compared against the original file.  If that original VCF contained complete genotypes, this will be used to calculate accuracy.',
                        style: 'padding-bottom: 10px;',
                        width: 1000
                    },{
                        xtype: 'checkbox',
                        fieldLabel: 'Skip Mendelian Check',
                        name: 'skipMendelianCheck'
                    },{
                        xtype: 'checkbox',
                        fieldLabel: 'Include Ancestors',
                        name: 'includeAncestors'
                    },{
                        xtype: 'ldk-integerfield',
                        fieldLabel: 'Dense Marker Batch Size',
                        name: 'denseMarkerBatchSize',
                        value: LABKEY.ActionURL.getParameter('denseMarkerBatchSize') || 2500
                    },{
                        xtype: 'textfield',
                        fieldLabel: 'Call Method',
                        name: 'callMethod',
                        value: LABKEY.ActionURL.getParameter('callMethod') || '2 0.8 0.9'
                    },{
                        xtype: 'ldk-integerfield',
                        fieldLabel: 'Min Genotype Quality',
                        name: 'minGenotypeQual',
                        value: LABKEY.ActionURL.getParameter('minGenotypeQual') || '20'
                    },{
                        xtype: 'ldk-integerfield',
                        fieldLabel: 'Min Genotype Depth',
                        name: 'minGenotypeDepth',
                        value: LABKEY.ActionURL.getParameter('minGenotypeDepth') || '10'
                    },{
                        xtype: 'checkbox',
                        fieldLabel: 'Split Each Set Into Separate Jobs',
                        itemId: 'splitSampleSets',
                        inputValue: true,
                        checked: !!LABKEY.ActionURL.getParameter('splitSampleSets') || true
                    },{
                        xtype: 'button',
                        border: true,
                        text: 'Add/Modify Groups',
                        handler: this.addSet,
                        scope: this
                    },{
                        xtype: 'panel',
                        style: 'padding-top: 5px;',
                        itemId: 'imputationSets'
                    }]
                }]);
            },

            addSet: function(btn){
                Ext4.create('Ext.window.Window', {
                    title: 'Add Set of Animals',
                    bodyStyle: 'padding: 5px;',
                    variantPanel: this,
                    border: false,
                    defaults: {
                        border: false
                    },
                    width: 800,
                    items: [{
                        html: 'In the box below you can paste sample Ids in 2 columns.  The first column contains a commas separated list of sample Ids where we should generate complete genotypes.  The second contains a list of sample Ids where we should generate only those matching the framework markers.  For example:<br><br>' +
                        '<pre>Sample1,Sample2\tSample3,Sample4</pre>',
                        style: 'padding-bottom: 10px;'
                    },{
                        xtype: 'textarea',
                        height: 300,
                        width: 780,
                        listeners: {
                            afterrender: function(field){
                                Ext4.defer(field.focus, 100, field);
                            }
                        }
                    }],
                    buttons: [{
                        text: 'Submit',
                        scope: this,
                        handler: function(btn){
                            var text = btn.up('window').down('textarea').getValue();
                            if (!text){
                                Ext4.Msg.alert('Error', 'Must enter a list of samples');
                                return;
                            }

                            text = Ext4.String.trim(text);
                            text = LDK.Utils.CSVToArray(text, '\t');

                            var win = btn.up('window');

                            var toAdd = [];
                            var errors = [];
                            Ext4.Array.forEach(text, function(line, idx){
                                if (line.length < 1){
                                    errors.push('Line ' + (idx+1) + ' does not have at least 2 elements');
                                    return;
                                }

                                toAdd.push([win.textToArray(line[0]), win.textToArray(line[1]), (line[2] ? win.textToArray(line[2]) : [])]);
                            }, this);

                            if (errors.length){
                                Ext4.Msg.alert('Error', errors.join('<br>'));
                                return;
                            }

                            win.addSamples(toAdd);
                        }
                    },{
                        text: 'Sample Helper',
                        scope: this,
                        handler: function(btn){
                            var win = btn.up('window');

                            Ext4.create('Ext.window.Window', {
                                width: 800,
                                title: 'Enter Sample Names',
                                bodyStyle: 'padding: 5px;',
                                items: [{
                                    html: 'This helper is designed to facilitate testing different combinations of WGS data, using the IDs listed below.  It produces a table w/ each set of WGS or GBS IDs, chosen from the appropriate VCF.  As reference data, it will preferentially use WGS as reference data, but select GBS if not available.  At minimum you need to provide the columns: SetName, GBS, and WGS.  You can optionally also provide the following columns, which will supercede the corresponding options above: frameworkFile, denseFile, alleleFrequencyFile, blacklistFile, callMethod.',
                                    border: false,
                                    style: 'padding-bottom: 10px;'
                                },{
                                    xtype: 'textarea',
                                    height: 300,
                                    width: 780
                                },{
                                    xtype: 'ldk-integerfield',
                                    itemId: 'wgsVcf',
                                    fieldLabel: 'Enter the VCF ID with WGS data',
                                    width: 300
                                },{
                                    xtype: 'ldk-integerfield',
                                    itemId: 'gbsVcf',
                                    fieldLabel: 'Enter the VCF ID with GBS data',
                                    width: 300
                                }],
                                addData: function(btn){
                                    var data = btn.up('window').down('textarea').getValue();
                                    data = Ext4.String.trim(data);

                                    var wgsVcf = btn.up('window').down('#wgsVcf').getValue();
                                    var gbsVcf = btn.up('window').down('#gbsVcf').getValue();

                                    if (!data || !wgsVcf || !gbsVcf){
                                        Ext4.Msg.alert('Error', 'Missing required fields');
                                        return false;
                                    }

                                    data = LDK.Utils.CSVToArray(data, '\t');

                                    var header = data.shift();

                                    //iterate rows
                                    var rows = [];
                                    Ext4.Array.forEach(data, function(row){
                                        var newRow = [];
                                        Ext4.Array.forEach(header, function(field, idx){
                                            var val = row[idx];
                                            if (field.toLowerCase() == 'wgs'){
                                                val = win.textToArray(val);
                                                Ext4.Array.forEach(val, function(r, i){
                                                    val[i] = wgsVcf + '||' + r;
                                                }, this);

                                                newRow[0] = val;
                                            }
                                            else if (field.toLowerCase() == 'gbs'){
                                                val = win.textToArray(val);

                                                newRow[1] = [];
                                                newRow[2] = [];
                                                Ext4.Array.forEach(val, function(r, i){
                                                    newRow[1].push(gbsVcf + '||' + r);

                                                    //determine if WGS VCF has this sample
                                                    if (win.variantPanel.sampleStore.findExact('sampleId', wgsVcf + '||' + r) > -1){
                                                        newRow[2].push(wgsVcf + '||' + r);
                                                    }
                                                    else {
                                                        newRow[2].push(gbsVcf + '||' + r);
                                                    }
                                                }, this);
                                            }
                                            else {
                                                if (!newRow[3]){
                                                    newRow[3] = {};
                                                }

                                                if (Ext4.isDefined(val)) {
                                                    newRow[3][field] = val;
                                                }
                                            }
                                        }, this);

                                        rows.push(newRow);
                                    }, this);

                                    win.addSamples(rows);

                                    return true;
                                },
                                buttons: [{
                                    text: 'Submit',
                                    scope: this,
                                    handler: function(btn){
                                        if (btn.up('window').addData(btn)) {
                                            btn.up('window').close();
                                        }
                                    }
                                },{
                                    text: 'Cancel',
                                    handler: function(btn){
                                        btn.up('window').close();
                                    }
                                }]
                            }).show();
                        }
                    },{
                        text: 'Close',
                        handler: function(btn){
                            btn.up('window').close();
                        }
                    }],
                    getAllCombinations: function (arr, reordered, currentIdx, level, size, ret){
                        if (level == size){
                            ret.push(Ext4.Array.clone(reordered));
                            return;
                        }

                        var arrays = [];
                        for (var i = currentIdx; i < arr.length; i++) {
                            reordered[level] = arr[i];
                            this.getAllCombinations(arr, reordered, i + 1, level + 1, size, ret);
                            if (i < arr.length-1 && arr[i] == arr[i+1]){
                                i++;
                            }
                        }
                    },
                    textToArray: function(val){
                        if (Ext4.isEmpty(val)){
                            return new Array();
                        }

                        val = val.replace(/[\s,;]+/g, ';');
                        val = val.replace(/(^;|;$)/g, '');
                        val = val.toLowerCase();

                        if(val)
                            val = val.split(';');
                        else
                            val = new Array();

                        return val;
                    },

                    getPropDisplay: function(obj){
                        var ret = [];
                        Ext4.Object.each(obj, function(key, value){
                            if (Ext4.isDefined(value)) {
                                ret.push(key + ': ' + value);
                            }
                        }, this);

                        return ret.join('<br>');
                    },

                    addSamples: function(toAdd){
                        this.close();

                        var target = this.variantPanel.down('#imputationSets');

                        var items = [];
                        Ext4.Array.forEach(toAdd, function(r){
                            items.push({
                                html: r[0].join(';<br>'),
                                maxWidth: 300
                            });

                            items.push({
                                html: r[1].join(';<br>'),
                                maxWidth: 300
                            });

                            items.push({
                                html: r[2] ? r[2].join(';<br>') : '',
                                maxWidth: 300
                            });

                            items.push({
                                html: r[3] ? this.getPropDisplay(r[3]) : '',
                                maxWidth: 300
                            });
                        }, this);

                        target.rawData = toAdd;
                        target.removeAll();
                        if (items.length) {
                            target.add({
                                layout: {
                                    type: 'table',
                                    columns: 4
                                },
                                defaults: {
                                    bodyStyle: 'padding: 5px;',
                                    border: 4,
                                    style: 'vertical-align:text-top;'
                                },
                                items: [{
                                    html: 'Complete Genotypes',
                                    maxWidth: 300
                                },{
                                    html: 'Imputed',
                                    maxWidth: 300
                                },{
                                    html: 'Reference',
                                    maxWidth: 300
                                },{
                                    html: 'Other Fields',
                                    maxWidth: 600
                                }].concat(items)
                            });
                        }
                    }
                }).show(btn);
            },

            onSubmit: function(){
                var vals = this.getForm().getFieldValues();
                var sampleSets = Ext4.Array.clone(this.down('#imputationSets').rawData);

                var jobName = this.down('#jobName').getValue();

                var errors = [];
                if (!vals.frameworkFile)
                    errors.push('Missing value for framework marker file');

                if (!vals.denseFile)
                    errors.push('Missing value for dense marker file');

                if (!sampleSets || Ext4.isEmpty(sampleSets))
                    errors.push('No sample sets provided');

                if (!vals.callMethod)
                    errors.push('Missing value for GIGI call method');

                if (errors.length){
                    Ext4.Msg.alert('Error', errors.join('<br>'));
                    return;
                }

                Ext4.Msg.wait('Saving...');
                if (!this.down('#splitSampleSets').getValue()){
                    vals.sampleSets = [];

                    Ext4.Array.forEach(sampleSets, function(ss){
                        while (ss.length > 3){
                            ss.pop();
                        }

                        vals.sampleSets.push(ss);
                    }, this);

                    if (!jobName){
                        Ext4.Msg.hide();
                        Ext4.Msg.alert('Error', 'Must provide a job name');
                        return;
                    }

                    LABKEY.Ajax.request({
                        url: LABKEY.ActionURL.buildURL('sequenceanalysis', 'runSequenceHandler'),
                        jsonData: {
                            handlerClass: 'org.labkey.variantdb.analysis.ImputationAnalysis',
                            outputFileIds: this.outputFileIds,
                            params: Ext4.encode(vals),
                            jobName: jobName
                        },
                        scope: this,
                        failure: LABKEY.Utils.getCallbackWrapper(LDK.Utils.getErrorCallback()),
                        success: function() {
                            Ext4.Msg.hide();

                            window.location = LABKEY.ActionURL.buildURL('pipeline-status', 'showList');
                        }
                    });
                }
                else {
                    var multi = new LABKEY.MultiRequest();
                    var success = true;
                    Ext4.Array.forEach(sampleSets, function(ss, idx){
                        var paramCopy = Ext4.apply({}, vals);
                        var jn = jobName;
                        if (ss.length > 3){
                            ss = Ext4.Array.clone(ss);
                            Ext4.Object.each(ss[3], function(key, value){
                                paramCopy[key] = value;

                                if (key.toLowerCase() == 'setname'){
                                    jn = value;
                                }
                            }, this);

                            ss.pop();
                        }
                        paramCopy.sampleSets = [ss];
                        console.log(paramCopy);

                        if (!jn){
                            Ext4.Msg.hide();
                            Ext4.Msg.alert('Error', 'Must provide a job name');
                            success = false;
                            return false;
                        }

                        multi.add(LABKEY.Ajax.request, {
                            url: LABKEY.ActionURL.buildURL('sequenceanalysis', 'runSequenceHandler'),
                            jsonData: {
                                handlerClass: 'org.labkey.variantdb.analysis.ImputationAnalysis',
                                outputFileIds: this.outputFileIds,
                                params: Ext4.encode(paramCopy),
                                jobName: jn
                            },
                            scope: this,
                            failure: LABKEY.Utils.getCallbackWrapper(LDK.Utils.getErrorCallback())
                        });
                    }, this);

                    if (success) {
                        multi.send(function () {
                            Ext4.Msg.hide();

                            window.location = LABKEY.ActionURL.buildURL('pipeline-status', 'showList');
                        }, this);
                    }
                }
            }
        });

        var outputFileIds = LABKEY.ActionURL.getParameter("outputFileIds");
        if (!outputFileIds){
            Ext4.Msg.alert('Error', 'Must provide output file IDs');
            return;
        }
        outputFileIds = outputFileIds.split(';');

        var webpart = <%=webpartContext%>;
        Ext4.create('VariantDB.panel.ImputationPanel', {
            outputFileIds: outputFileIds
        }).render(webpart.wrapperDivId);
    });

</script>