<h4>Release 2.2:</h4>
<ul>
    <li>This is an additional 103 animals over the prior version.</li>
    <li>We made two changes to how variants are filters:</li>
    <ol>
        <li>We now use the official NCBI RepeatMasker data to mask variants in repetitive regions, instead of a custom RepeatMasker file generated in-house for version 2.1 and lower. In practice the result should be very similar.</li>
        <li>We removed the SnpCluster filter from processing. In previous versions, any SNPs falling into a window with at least 3 SNPs in a 10bp window were filtered. As the dataset grew, this captured too many valid variants. This change resulted in a sizable increase in passing variants in this release</li>
    </ol>
    <li>We upgraded the mGAP website to use JBrowse 2, a faster, more powerful version of the JBrowse 1 browser used in prior versions. Please use the help link to ask any questions or report issues.</li>
</ul>

<h4>Release 2.1:</h4>
<ul>
    <li>This is an iterative improvement to the dataset, adding 275 additional animals.</li>
    <li>We removed 2 animals (m05247 and m03764), which we believe are poor quality/contaminated samples.</li>
    <li>There was an update to how all whole exome data is processed, applied across all data. We now run MarkDuplicates on the BAM file, to mark/ignore duplicate reads prior to variant calling. This should also remove some previous artifacts from the dataset.</li>
</ul>

<h4>Release 2.0:</h4>
<ul>
    <li>Substantial revamp of all data. All samples have been realigned to the <a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_003339765.1/">MMul_10 reference genome</a>, followed by our <a href="mgap-dataProcessing.view">standard GenotypeGVCFs pipeline</a>. The MMul_10 is the most complete rhesus macaque assembly to date, and we expect this should improve accuracy of variant calls. Further, because our data are now aligned to the same assembly as NCBI/Ensembl, it should be easier to translate between mGAP and other databases.</li>
    <li>Our internal variant calling process has switched to use GATK's GenomicsDB to pre-aggregate data prior to calling with GenotypeGVCFs, as opposed to CombineGVCFs, which was used in prior releases. This should be a purely technical difference with no change in the resulting data</li>
</ul>

<h4>Future Plans:</h4>
<ul>
    <li>We will support other modes of viewing and downloading variant data, in particular tabular views by gene.</li>
    <li>We recognize that the mGAP release VCF can be enormous, particularly because of all the site-specific functional annotation. To support different types of users, upcoming releases will include 'slim' versions of the data, which will be downloadable files with certain information removed to save file size.</li>
</ul>